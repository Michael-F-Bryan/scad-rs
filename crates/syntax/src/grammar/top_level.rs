use crate::{
    grammar::{expressions, TokenSet},
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const CONTINUE: TokenSet = TokenSet::new([T![;], T!["}"]]);

pub(crate) fn package(p: &mut Parser<'_>) {
    let m = p.start();

    while !p.at(EOF) {
        statement(p);
    }

    p.complete(m, PACKAGE);
}

pub(crate) fn statement(p: &mut Parser<'_>) {
    let m = p.start();

    match p.current() {
        T![include] => include(p, m),
        T![use] => use_(p, m),
        IDENT if p.nth_at(1, T![=]) => {
            assignment(p, m);
            p.expect(T![;]);
        }
        _ => p.error_recover("Expected a statement", m, CONTINUE),
    }
}

pub(crate) fn include(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![include]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, INCLUDE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn use_(p: &mut Parser<'_>, m: Mark) {
    p.bump(T![use]);

    if p.at(STRING) {
        p.bump(STRING);
        p.complete(m, USE);
    } else {
        p.error_recover("Expected a file name", m, CONTINUE);
    }
}

pub(crate) fn assignment(p: &mut Parser<'_>, m: Mark) {
    p.bump(IDENT);

    p.eat(T![=]);

    expressions::expr(p);
    p.complete(m, ASSIGNMENT);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        empty_package: package(""),
        include_statement: statement(r#"include "./foo/bar""#),
        use_statement: statement(r#"use "./foo/bar""#),
        assignment_statement: statement("x = 42;"),
        kitchen_sink: package(r#"
            include "foo/bar.scad"
            use "baz.scad"
            a = 42;
            b = (42);
            c = 1 + 1;
        "#),
    }
}
