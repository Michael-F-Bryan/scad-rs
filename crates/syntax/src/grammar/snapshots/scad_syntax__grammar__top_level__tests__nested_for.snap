---
source: crates/syntax/src/grammar/top_level.rs
description: "statement(for(i = [0: 10], j = [0, 360]) cube(i, j);)"
---
Node(
    FOR_STATEMENT@0..42
      FOR_KW@0..3 "for"
      L_PAREN@3..4 "("
      ASSIGNMENTS@4..29
        ASSIGNMENT@4..15
          IDENT@4..5 "i"
          WHITESPACE@5..6 " "
          EQUALS@6..7 "="
          RANGE_EXPR_FROM_TO@7..15
            WHITESPACE@7..8 " "
            L_BRACKET@8..9 "["
            INTEGER@9..10
              INTEGER@9..10 "0"
            COLON@10..11 ":"
            INTEGER@11..14
              WHITESPACE@11..12 " "
              INTEGER@12..14 "10"
            R_BRACKET@14..15 "]"
        COMMA@15..16 ","
        ASSIGNMENT@16..29
          WHITESPACE@16..17 " "
          IDENT@17..18 "j"
          WHITESPACE@18..19 " "
          EQUALS@19..20 "="
          LIST_EXPR@20..29
            WHITESPACE@20..21 " "
            L_BRACKET@21..22 "["
            INTEGER@22..23
              INTEGER@22..23 "0"
            COMMA@23..24 ","
            INTEGER@24..28
              WHITESPACE@24..25 " "
              INTEGER@25..28 "360"
            R_BRACKET@28..29 "]"
      R_PAREN@29..30 ")"
      MODULE_INSTANTIATION@30..42
        WHITESPACE@30..31 " "
        IDENT@31..35 "cube"
        L_PAREN@35..36 "("
        ARGUMENTS@36..40
          LOOKUP_EXPR@36..37
            IDENT@36..37 "i"
          COMMA@37..38 ","
          LOOKUP_EXPR@38..40
            WHITESPACE@38..39 " "
            IDENT@39..40 "j"
        R_PAREN@40..41 ")"
        SEMICOLON@41..42
          SEMICOLON@41..42 ";"
    ,
)
